flink-core/src/main/java/org/apache/flink/api/connector/source/lib/util/GuavaRateLimiter.java

```diff
@@ -0,0 +1,38 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.api.connector.source.lib.util;
+
+import org.apache.flink.shaded.guava30.com.google.common.util.concurrent.RateLimiter;
+
+/** An implementation of {@link RateLimiter} based on Guava's RateLimiter. */
+public class GuavaRateLimiter
```

Comment:
Use `GuavaFlinkConnectorRateLimiter` instead (or unify them). We shouldn't have 2 different rate limiters within flink-core for connectors.

Reply:
`GuavaFlinkConnectorRateLimiter` depends on the `RuntimeContext`, which we do not expose in the new Source interface model, so unifying them won't work. `FlinkConnectorRateLimiter` is only used in the `PubSubSource`, which is based on the old source interfaces, has to be rewritten, and will probably be externalized out of the main Flink project. I think we should consider those old rate-limiting interfaces as deprecated.

Reply:
You could add a new open() method to the `GuavaFlinkConnectorRateLimiter` that accepts an int, and add a default implementation to the existing open() method that calls `open(runtimeContext#getParallelism())`

Reply:
@zentol Since we converged on the design of the `RateLimiter` interface with checkpoint awareness, would you still prefer to merge the two?

Reply:
let's merge it as is and resolve the duplication in a follow-up.

The FlinkConnectorRateLimiter interface is a bit of a problem; it is good that it allows multiple permits to be acquired at once for efficiency purposes, but it interacts oddly with rate limiting.
For example, if your limit is 10 records per checkpoint, then you're never getting any permits if you request 11 at once.

We'll need to think about how we can properly implement this.